Make directory permissions configurable

hoegaarden opened the pull request:
- Inner directories inherit permissions from the outer directory
- the permissions on the directories is changed late, so that pulling in content is not blocked

/xref: https://github.com/vmware-tanzu/carvel-vendir/issues/186

hoegaarden committed:
[dirPerms] Get test infra in place

hoegaarden committed:
[dirPerms] Make permissions of inner and outer directories configurable

hoegaarden committed:
[dirPerms] Inner dirs inherit permissions from the outer one

cppforlife commented:
cc @joaopapereira you previously had some thoughts on permissions. any objections to this PR?

joaopapereira commented on test/e2e/directory_permissions_test.go:
```diff
@@ -0,0 +1,171 @@
+
+	"github.com/stretchr/testify/assert"
+	"github.com/vmware-tanzu/carvel-vendir/pkg/vendir/config"
+	"gopkg.in/yaml.v2"
```
We are using this library inside the code, maybe it makes sense to keep it consistent. I just realized that the gopkg one is used in a different tests, I will create a PR to fix that as well
```suggestion
	"sigs.k8s.io/yaml"
```

joaopapereira commented on test/e2e/directory_permissions_test.go:
```diff
@@ -0,0 +1,171 @@
+
+func writeConfigFile(t *testing.T, tmpDir string, config config.Config) {
+	vendirYML, err := os.Create(filepath.Join(tmpDir, "vendir.yml"))
+	assert.NoError(t, err, "creating vendir.yml")
```
I think we should use require in this place as well as in other parts. The main difference between `require` and `assert` is that assert allows the these to continue, but in situations like this we should not continue with the execution of the test. (I did not highlight all the places where this change can be done)
```suggestion
	require.NoError(t, err, "creating vendir.yml")
```

joaopapereira reviewed (changes requested):
Hey, thanks for putting together this PR.

The PR implementation looks good, I added a couple of comments mainly on the test side of the world.
Also when you finish this do you mind creating a PR in https://github.com/vmware-tanzu/carvel to update the documentation with the new permission field? The website is organized by version and your changes should be inside [this directory](https://github.com/vmware-tanzu/carvel/tree/develop/site/content/vendir/docs/develop)

joaopapereira reviewed (approved):
LGTM

cppforlife reviewed (commented):
lets rename field name to be plural: permission => permissions

hoegaarden committed:
[dirPerms] Rename the permissions field

Use `Permissions` instead of `Permission` in the Directory and
DirectoryContents config struct; and similarily `permissions` instead of
`permission` in the serialized config.
